reference-delegate-explainer.md

```diff
@@ -0,0 +1,534 @@
+   * `<combobox-listbox>` allows the target of its `aria-activedescendant` to be controlled externally via its custom `activeitem` attribute.
+
+#### `<combobox-input>`
+
+This component is a wrapper around an `<input>` for use by a combobox.
+
+1. It delegates all references to the input. This lets, for example, a label for this component to be applied to the input.
+2. A custom attribute `listbox` is hooked up to both `ariaControlsElements` and `ariaActiveDescendantElement`.
+   * The listbox delegates the two attributes to different elements inside (see `<combobox-listbox>` below), but this component references the parent listbox for both.
+
+```html
+<template id="t-combobox-input">
+  <input id="real-input" role="combobox" type="text" />
+</template>
+
+<script>
+customElements.define("combobox-input", 
+  class ComboboxInput extends HTMLElement {
+    static observedAttributes = ["listbox"];
+
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-combobox-input").content.cloneNode(true));
+      this.shadowRoot_.referenceDelegate = "real-input"; // (1)
+      this.input_ = this.shadowRoot_.getElementById('real-input');
+    }
+
+    attributeChangedCallback(attr, _oldValue, value) {
+      if (attr === "listbox") {
+        // (2)
+        const listbox = this.getRootNode().getElementById(value);
+        this.input_.ariaControlsElements = [listbox];
+        this.input_.ariaActiveDescendantElement = listbox;
+      }
+    }
+  });
+</script>
+```
+
+#### `<combobox-listbox>`
+
+This component is a wrapper around `<div role="listbox">` and the `<div role="option">` items inside.
+
+1. It delegates all references _except_ `aria-activedescendant` to the `<div role="listbox">` using `referenceDelegate`.
+2. It has a custom attribute `activeitem`, which is used to control which item gets the `aria-activedescendant` delegation using `referenceDelegateMap`. This lets the parent component control the active item.
+
+```html
+<template id="t-combobox-listbox">
+  <div id="real-listbox" role="listbox">
+    <div id="option-1" role="option">Option 1</div>
+    <div id="option-2" role="option">Option 2</div>
+  </div>
+</template>
+
+<script>
+customElements.define("combobox-listbox", 
+  class ComboboxListbox extends HTMLElement {
+    static observedAttributes = ["activeitem"];
+
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-combobox-listbox").content.cloneNode(true));
+      this.shadowRoot_.referenceDelegate = "real-listbox"; // (1)
+    }
+
+    attributeChangedCallback(attr, _oldValue, value) {
+      if (attr === "activeitem") {
+        this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = value; // (2)
+      }
+    }
```

Comment:
Minor quibble: I think a listbox component would provide better encapsulation by exposing methods to move the focus up/down rather than an `attributeChangedCallback` explicitly setting the index of the active item. The issue here is that the consumer of the listbox has to "know" how many items are in the listbox and which one is the active one (thus exposing internal details).

Whereas using methods (or events or what have you), the fancy input could just communicate to the listbox "Hey, the <kbd>↑</kbd> key was pressed," so the listbox could change its `referenceDelegateMap` internally without the consumer knowing how many items there are or which one is the active one.

Reply:
That does make sense. However, my goal was more to show how the referenceTargetMap works, rather than design a real listbox/combox. I'm inclined to leave it as-is to keep things simpler for the example.